imports.go

```diff
@@ -1,142 +1,199 @@
-	cwd, err := os.Getwd()
-	if err != nil {
-		return nil, err
-	}
+func filterPackages(packages []string) []string {
+	var imports []string
 
-	var (
-		allImports = map[string]bool{}
-		imports    = []string{}
-	)
+	for _, importing := range packages {
+		importpath, err := getRootImportpath(importing)
+		if err != nil {
+			continue
+		}
 
-	filepath.Walk(
-		cwd, func(path string, info os.FileInfo, err error) error {
-			if err != nil {
-				return err
-			}
+		if inTests {
+			importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
+		}
 
-			if filepath.Base(path) == ".git" ||
-				filepath.Dir(path) == filepath.Join(cwd, "vendor") {
-				return filepath.SkipDir
-			}
+		if isOwnPackage(importpath) {
+			continue
+		}
 
-			if path == filepath.Join(cwd, "vendor") {
-				return nil
-			}
+		// Skip anything that is already vendored upstream
+		if strings.Contains(importpath, "/vendor/") {
+			continue
+		}
 
-			if !info.IsDir() {
-				return nil
+		found := false
+		for _, imported := range imports {
+			if importpath == imported {
+				found = true
+				break
 			}
+		}
 
-			err = recursiveParseImports(
-				allImports,
-				".",
-				path,
-			)
-			if _, ok := err.(*build.NoGoError); ok {
-				return nil
-			}
+		if found {
+			continue
+		}
 
-			if err != nil {
-				log.Println(err)
-			}
+		imports = append(imports, importpath)
+	}
 
-			return nil
-		},
-	)
+	return imports
+}
 
-	for importing, standard := range allImports {
-		if !standard {
-			importpath, err := getRootImportpath(importing)
-			if err != nil {
-				continue
-			}
+func ensureDependenciesExist(packages []string, includeTestingDependencies bool) error {
+	args := []string{"get"}
+	if includeTestingDependencies {
+		args = append(args, "-t")
+	}
+	args = append(args, packages...)
 
-			if inTests {
-				importpath = strings.Replace(importpath, "__blankd__", "localhost:60001", -1)
-			}
+	out, err := execute(exec.Command("go", args...))
+	if err != nil {
+		return fmt.Errorf("couldn't 'go get' dependencies:\n%s", out)
+	}
 
-			if isOwnPackage(importpath, cwd) {
-				continue
-			}
+	return nil
+}
 
-			found := false
-			for _, imported := range imports {
-				if importpath == imported {
-					found = true
-					break
-				}
-			}
+func list(pkg string) (listOutput, error) {
+	data := listOutput{}
 
-			if found {
-				continue
-			}
+	out, err := execute(exec.Command("go", "list", "-e", "-json", pkg))
+	if err != nil {
+		return data, err
+	}
 
-			imports = append(imports, importpath)
-		}
+	err = json.Unmarshal([]byte(out), &data)
+	if err != nil {
+		return data, fmt.Errorf("can't unmarshal go list JSON output: %s\n%q", err, out)
 	}
 
-	sort.Strings(imports)
+	return data, nil
+}
 
-	return imports, nil
+func listPackages() ([]string, error) {
+	var packages []string
+
+	out, err := execute(exec.Command("go", "list", "-e", "./..."))
+	if err != nil {
+		return packages, err
+	}
+
+	r := strings.NewReader(out)
+	s := bufio.NewScanner(r)
```

Comment:
wow, reader and scanner, may be just use strings.Split(strings.TrimSpaces(out), "\n")?